From 3bf2d35d66c16ec531d018790d5f78d159878435 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 11 Sep 2014 22:43:50 -0400
Subject: [infiniband] cxgb4: log detailed warnings for negative advice

Message-id: <1410475447-94392-92-git-send-email-svemuri@redhat.com>
Patchwork-id: 93855
O-Subject: [RHEL7.1 PATCH BZ 1124947 091/108] iw_cxgb4: log detailed warnings for negative advice
Bugzilla: 1124947
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Signed-off-by: Steve Wise <swise@opengridcomputing.com>
Signed-off-by: Hariprasad Shenai <hariprasad@chelsio.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit dd92b12453b34850912ccdeefa740a2c96f870c2)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/cxgb4/cm.c b/drivers/infiniband/hw/cxgb4/cm.c
index addebc1..0ad68f6 100644
--- a/drivers/infiniband/hw/cxgb4/cm.c
+++ b/drivers/infiniband/hw/cxgb4/cm.c
@@ -1822,6 +1822,20 @@ static int is_neg_adv(unsigned int status)
         status == CPL_ERR_KEEPALV_NEG_ADVICE;
 }
 
+static char *neg_adv_str(unsigned int status)
+{
+ switch (status) {
+ case CPL_ERR_RTX_NEG_ADVICE:
+  return "Retransmit timeout";
+ case CPL_ERR_PERSIST_NEG_ADVICE:
+  return "Persist timeout";
+ case CPL_ERR_KEEPALV_NEG_ADVICE:
+  return "Keepalive timeout";
+ default:
+  return "Unknown";
+ }
+}
+
 static void set_tcp_window(struct c4iw_ep *ep, struct port_info *pi)
 {
  ep->snd_win = snd_win;
@@ -2019,8 +2033,9 @@ static int act_open_rpl(struct c4iw_dev *dev, struct sk_buff *skb)
       status, status2errno(status));
 
  if (is_neg_adv(status)) {
-  printk(KERN_WARNING MOD "Connection problems for atid %u\n",
-   atid);
+  dev_warn(&dev->rdev.lldi.pdev->dev,
+    "Connection problems for atid %u status %u (%s)\n",
+    atid, status, neg_adv_str(status));
   return 0;
  }
 
@@ -2495,8 +2510,9 @@ static int peer_abort(struct c4iw_dev *dev, struct sk_buff *skb)
 
  ep = lookup_tid(t, tid);
  if (is_neg_adv(req->status)) {
-  PDBG("%s neg_adv_abort ep %p tid %u\n", __func__, ep,
-       ep->hwtid);
+  dev_warn(&dev->rdev.lldi.pdev->dev,
+    "Negative advice on abort - tid %u status %d (%s)\n",
+    ep->hwtid, req->status, neg_adv_str(req->status));
   return 0;
  }
  PDBG("%s ep %p tid %u state %u\n", __func__, ep, ep->hwtid,
@@ -3901,8 +3917,9 @@ static int peer_abort_intr(struct c4iw_dev *dev, struct sk_buff *skb)
   return 0;
  }
  if (is_neg_adv(req->status)) {
-  PDBG("%s neg_adv_abort ep %p tid %u\n", __func__, ep,
-       ep->hwtid);
+  dev_warn(&dev->rdev.lldi.pdev->dev,
+    "Negative advice on abort - tid %u status %d (%s)\n",
+    ep->hwtid, req->status, neg_adv_str(req->status));
   kfree_skb(skb);
   return 0;
  }
-- 
1.7.1